### PR TITLE
OCC-35: Tax Abstraction

### LINKED ISSUE #157: Tax abstraction
Part of #33

## Content Parts

`TaxPart`
- `TextField ProductTaxCode`: a product or product category ID for taxing purposes (can be used to indicate which VAT rate applies on domestic sales).
- `Amount GrossPrice`: if specified, then this is used instead of `ProductPart.Price`.
- `NumericField GrossPriceRate`: if specified, then this is used to get the net price from the `GrossPrice`.

All fields are optional. Either both `GrossPrice` and `GrossPriceRate` should be filled or neither.
A handler should detect when both `GrossPrice` and `GrossPriceRate` are filled, calculate and update the correct `PricePart.Price` as the net price.

## Models and Services

- Add a non-serializing gross price property to `ShoppingCartItem` or `PrioritizedPrice` that the following services can use to store calculated tax.
- Add `ITaxProvider` and `ITaxService`. Similar structure to `IPriceProvider` and `IPriceService` respectively, just referring to tax.

## Front-end
- Prices on the front-end should check for `TaxPart.GrossPrice` and if not present (so using net price) then display "{price} +tax {currency}" for clarity.

### PR DESCRIPTION
**Pull request by DAud-IcI**
OCC-35

Fixes #157

### CODE REVIEW

File: src/Modules/OrchardCore.Commerce.Tax/OrchardCore.Commerce.Tax.csproj [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    <None Include="License.md" Pack="true" PackagePath="" />
+    <None Include="Readme.md" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\OrchardCore.Commerce.ContentFields\OrchardCore.Commerce.ContentFields.csproj" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Folder Include="Handlers" />
```
Comment (DemeSzabolcs) [inline src/Modules/OrchardCore.Commerce.Tax/OrchardCore.Commerce.Tax.csproj:39]: Is this necessary?

**Review by DemeSzabolcs — CHANGES_REQUESTED** [review]

File: src/Modules/OrchardCore.Commerce.Tax/Migrations/TaxPartMigrations.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                    .WithSettings(new TextFieldSettings
+                    {
+                        Hint = "The ID that identifies the product of product category for taxing purposes.",
+                    }))
+                .WithField(part => part.GrossPrice, part => part
+                    .WithDisplayName("Gross Price")
+                    .WithSettings(new PriceFieldSettings
+                    {
+                        Hint = "The price with tax. If specified along with the Tax Rate, then Price content part is " +
+                               "updated to the calculated net value on publish..",
```
[the remark this answers is not in the document]
Comment (DAud-IcI) [reply src/Modules/OrchardCore.Commerce.Tax/Migrations/TaxPartMigrations.cs:39]:
```suggestion
                               "updated to the calculated net value on publish.",
```

File: src/Modules/OrchardCore.Commerce/Drivers/OrderPartDisplayDriver.cs [hunk 3]
```diff
@@ -90,24 +93,33 @@ private async ValueTask PopulateViewModelAsync(OrderPartViewModel model, OrderPa
...
-                model.LineItems.Add(item);
+            taxContext = await _taxProviders.UpdateWithFirstApplicableProviderAsync(taxContext);
+            total = taxContext.TotalsByCurrency.Single();
 
-                total += item.LinePrice;
+            foreach (var (item, index) in taxContext.Items.Select((item, index) => (item, index)))
+            {
+                var lineItem = lineItems[index];
+                lineItem.LinePrice = item.UnitPrice;
+                lineItem.UnitPrice = item.Subtotal;
```
[the remark this answers is not in the document]
Comment (DemeSzabolcs) [reply src/Modules/OrchardCore.Commerce/Drivers/OrderPartDisplayDriver.cs]: The order is fixed:
image
But I just noticed that it is also wrong on the shopping cart's page: 
image

[the remark this answers is not in the document]
Comment (DemeSzabolcs) [reply src/Modules/OrchardCore.Commerce/Drivers/OrderPartDisplayDriver.cs]: The order is fixed:
image
But I just noticed that it is also wrong on the shopping cart's page: 
image

**Review by DemeSzabolcs — CHANGES_REQUESTED** [review]

**Pull request merged**